pkg/alibabacloud/eni/node.go

```diff
@@ -0,0 +1,436 @@
+			fieldENIID:     e.NetworkInterfaceID,
+			"numAddresses": len(e.PrivateIPSets),
+		}).Debug("Considering ENI for IP release")
+
+		// Count free IP addresses on this ENI
+		ipsOnENI := n.k8sObj.Status.AlibabaCloud.ENIs[e.NetworkInterfaceID].PrivateIPSets
+		freeIpsOnENI := []string{}
+		for _, ip := range ipsOnENI {
+			// exclude primary IPs
+			if ip.Primary {
+				continue
+			}
+			_, ipUsed := n.k8sObj.Status.IPAM.Used[ip.PrivateIpAddress]
+			if !ipUsed {
+				freeIpsOnENI = append(freeIpsOnENI, ip.PrivateIpAddress)
+			}
+		}
+		freeOnENICount := len(freeIpsOnENI)
+
+		if freeOnENICount <= 0 {
+			continue
+		}
+
+		scopedLog.WithFields(logrus.Fields{
+			fieldENIID:       e.NetworkInterfaceID,
+			"excessIPs":      excessIPs,
+			"freeOnENICount": freeOnENICount,
+		}).Debug("ENI has unused IPs that can be released")
+		maxReleaseOnENI := math.IntMin(freeOnENICount, excessIPs)
+
+		r.InterfaceID = key
+		r.PoolID = ipamTypes.PoolID(e.VPC.VPCID)
+		r.IPsToRelease = freeIpsOnENI[:maxReleaseOnENI]
+	}
+
+	return r
+}
+
+// ReleaseIPs performs the ENI IP release operation
+func (n *Node) ReleaseIPs(ctx context.Context, r *ipam.ReleaseAction) error {
+	return n.manager.api.UnassignPrivateIPAddresses(ctx, r.InterfaceID, r.IPsToRelease)
+}
+
+// GetMaximumAllocatableIPv4 returns the maximum amount of IPv4 addresses
+// that can be allocated to the instance
+func (n *Node) GetMaximumAllocatableIPv4() int {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+
+	// Retrieve l for the instance type
+	l, limitsAvailable := n.getLimitsLocked()
+	if !limitsAvailable {
+		return 0
+	}
+
+	// Return the maximum amount of IP addresses allocatable on the instance
+	// reserve Primary eni
+	return (l.Adapters - 1) * l.IPv4
+}
+
+func (n *Node) loggerLocked() *logrus.Entry {
+	if n == nil || n.instanceID == "" {
+		return log
+	}
+
+	return log.WithField("instanceID", n.instanceID)
+}
+
+// getLimits returns the interface and IP limits of this node
+func (n *Node) getLimits() (ipamTypes.Limits, bool) {
+	n.mutex.RLock()
+	l, b := n.getLimitsLocked()
+	n.mutex.RUnlock()
+	return l, b
+}
+
+// getLimitsLocked is the same function as getLimits, but assumes the n.mutex
+// is read locked.
+func (n *Node) getLimitsLocked() (ipamTypes.Limits, bool) {
+	return limits.Get(n.k8sObj.Spec.AlibabaCloud.InstanceType)
+}
+
+func (n *Node) getSecurityGroupIDs(ctx context.Context, eniSpec eniTypes.Spec) ([]string, error) {
+	// 1. use security group defined by user
+	// 2. use security group used by eth0
+
+	if len(eniSpec.SecurityGroups) > 0 {
+		return eniSpec.SecurityGroups, nil
+	}
+
+	if len(eniSpec.SecurityGroupTags) > 0 {
+		securityGroups := n.manager.FindSecurityGroupByTags(eniSpec.VPCID, eniSpec.SecurityGroupTags)
+		if len(securityGroups) == 0 {
+			n.loggerLocked().WithFields(logrus.Fields{
+				"vpcID": eniSpec.VPCID,
+				"tags":  eniSpec.SecurityGroupTags,
+			}).Warn("No security groups match required vpc id and tags, using eth0 security groups")
```

Comment:
```suggestion
			}).Warn("No security groups match required VPC ID and tags, using eth0 security groups")
```

Could we also clarify what is meant by "eth0 security groups"? A regular use won't understand what that means. FWIW, I I'm not sure what it means either :)

Reply:
For what it's worth, this seems to be present in the original code as well, and I find it equally confusing ;)

Reply:
ENI must have at least one security group, this is a fallback to retreat primary ENI's security group.
I'm going to change to the `No security groups match required VPC ID and tags, using primary ENI's security groups"`